src/manager/checkpoint/manager/bottomup.rs

```diff
@@ -1,49 +1,288 @@
+            .ok_or_else(|| anyhow!("chain head has empty cid: {:}", self.child_subnet.id))?;
+        checkpoint.data.proof = Cid::try_from(child_tip_set)?.to_bytes();
+
+        let to = self.child_subnet.id.subnet_actor();
+        let message = MpoolPushMessage::new(
+            to,
+            *validator,
+            ipc_subnet_actor::Method::SubmitCheckpoint as MethodNum,
+            cbor::serialize(&checkpoint, "checkpoint")?.to_vec(),
+        );
+        let mem_push_response = self
+            .parent_client
+            .mpool_push_message(message)
+            .await
+            .map_err(|e| {
+                log::error!(
+                    "error submitting checkpoint for epoch {epoch:} in subnet: {:?}",
+                    self.child_subnet.id
+                );
+                e
+            })?;
+
+        // wait for the checkpoint to be committed before moving on.
+        let message_cid = mem_push_response.cid()?;
+        log::debug!("checkpoint message published with cid: {message_cid:?}");
+
+        Ok(())
     }
 
     async fn next_submission_epoch(
         &self,
-        _validator: &Address,
+        validator: &Address,
     ) -> anyhow::Result<Option<ChainEpoch>> {
-        todo!()
+        log::debug!(
+            "attempt to obtain the next submission epoch in bottom up checkpoint for subnet: {:?}",
+            self.child_subnet.id
+        );
+
+        let current_epoch = self.child_client.current_epoch().await?;
+        let last_executed_epoch = self.last_executed_epoch().await?;
+
+        log::debug!(
+            "latest epoch {:?}, last executed epoch: {:?} for bottom up checkpointing in subnet: {:?}",
+            current_epoch,
+            last_executed_epoch,
+            self.child_subnet.id,
+        );
+
+        let next_submission_epoch = last_executed_epoch + self.checkpoint_period;
+        if current_epoch < next_submission_epoch {
+            log::info!("latest epoch {current_epoch:} lagging next submission epoch {next_submission_epoch:}");
+            return Ok(None);
+        }
+
+        if self
+            .parent_client
+            .ipc_has_voted_bu_in_epoch(validator, &self.child_subnet.id, next_submission_epoch)
+            .await?
+        {
+            log::info!("next submission epoch {next_submission_epoch:} already executed");
+            return Ok(None);
+        }
+
+        log::debug!("next submission epoch {next_submission_epoch:}");
+
+        Ok(Some(next_submission_epoch))
     }
 }
+
+async fn obtain_checkpoint_period(
+    subnet_id: &SubnetID,
+    parent_client: &DefaultLotusJsonRPCClient,
+) -> anyhow::Result<ChainEpoch> {
+    log::debug!("Getting the bottom up checkpoint period for subnet: {subnet_id:?}");
+
+    // Read the parent's chain head and obtain the tip set CID.
+    let parent_head = parent_client.chain_head().await?;
+    let cid_map = parent_head.cids.first().unwrap().clone();
+    let parent_tip_set = Cid::try_from(cid_map)?;
```

Comment:
Same here, is it worth adding the tipset as an argument?

Reply:
see comment above